CombinatorialGames/SignExpansion/Basic.lean

```diff
@@ -0,0 +1,219 @@
+  isUpperSet_sign_preimage_singleton_zero := by
+    intro a b hab ha
+    rw [← WithTop.coe_le_coe] at hab
+    simp only [sign_eq_coe, Set.mem_preimage, Set.mem_singleton_iff, ite_eq_right_iff,
+      apply_eq_zero] at ha ⊢
+    exact fun hb => (ha (hab.trans_lt hb)).trans hab
+
+@[simp]
+theorem length_restrict (x : SignExpansion.{u}) (o : WithTop NatOrdinal.{u}) :
+    (x.restrict o).length = min x.length o := by
+  apply eq_of_forall_ge_iff
+  intro c
+  cases c with
+  | top => simp
+  | coe c =>
+    rw [← apply_eq_zero, restrict]
+    simp [imp_iff_or_not, apply_eq_zero]
+
+theorem restrict_apply_of_coe_lt {x : SignExpansion.{u}} {o₁ : WithTop NatOrdinal.{u}}
+    {o₂ : NatOrdinal.{u}} (h : o₂ < o₁) : x.restrict o₁ o₂ = x o₂ := by
+  simp [restrict, h]
+
+theorem restrict_apply_of_le_coe {x : SignExpansion.{u}} {o₁ : WithTop NatOrdinal.{u}}
+    {o₂ : NatOrdinal.{u}} (h : o₁ ≤ o₂) : x.restrict o₁ o₂ = 0 := by
+  simp [restrict, h]
+
+@[simp]
+theorem restrict_of_length_le {e : SignExpansion.{u}} {o : WithTop NatOrdinal.{u}}
+    (ho : e.length ≤ o) : e.restrict o = e := by
+  ext o'
+  obtain ho' | ho' := lt_or_ge (.some o') e.length
+  · rw [restrict_apply_of_coe_lt (ho'.trans_le ho)]
+  · rw [apply_of_length_le ho']
+    apply apply_of_length_le
+    simp [ho']
+
+theorem restrict_top_right {e : SignExpansion.{u}} : e.restrict ⊤ = e := by simp
+
+instance : Zero SignExpansion.{u} where
+  zero := {
+    sign _ := 0
+    isUpperSet_sign_preimage_singleton_zero := by simp [isUpperSet_univ]
+  }
+
+@[simp]
+theorem zero_apply (o : NatOrdinal.{u}) : (0 : SignExpansion.{u}) o = 0 := rfl
+
+theorem restrict_zero (e : SignExpansion.{u}) : e.restrict 0 = 0 := by
+  ext; simp [apply_eq_zero]
+
+@[simp]
+theorem length_zero : length 0 = 0 := by
+  rw [← restrict_zero 0, length_restrict, ← WithTop.coe_zero,
+    ← NatOrdinal.bot_eq_zero, WithTop.coe_bot, min_bot_right]
+
+instance linearOrder : LinearOrder SignExpansion.{u} :=
+  LinearOrder.lift' (toLex ⇑·) (by simp [Function.Injective])
+
+theorem le_iff_toLex {a b : SignExpansion.{u}} : a ≤ b ↔ toLex ⇑a ≤ toLex ⇑b := Iff.rfl
+theorem lt_iff_toLex {a b : SignExpansion.{u}} : a < b ↔ toLex ⇑a < toLex ⇑b := Iff.rfl
+
+instance : Bot SignExpansion.{u} where
+  bot := {
+    sign _ := -1
+    isUpperSet_sign_preimage_singleton_zero := by simp [isUpperSet_empty]
+  }
+
+@[simp]
+theorem bot_apply (o : NatOrdinal.{u}) : (⊥ : SignExpansion.{u}) o = -1 := rfl
+
+instance : Top SignExpansion.{u} where
+  top := {
+    sign _ := 1
+    isUpperSet_sign_preimage_singleton_zero := by simp [isUpperSet_empty]
+  }
+
+@[simp]
+theorem top_apply (o : NatOrdinal.{u}) : (⊤ : SignExpansion.{u}) o = 1 := rfl
+
+instance boundedOrder : BoundedOrder SignExpansion.{u} where
+  le_top _ := le_iff_toLex.2 <| Pi.toLex_monotone (by simp [Pi.le_def])
+  bot_le _ := le_iff_toLex.2 <| Pi.toLex_monotone (by simp [Pi.le_def])
+
+instance biheytingAlgebra : BiheytingAlgebra SignExpansion.{u} :=
```

Comment:
Do we actually need to use anything from this instance? I think we can omit this and declare it when we get the full complete linear order instance.